transport/http/agent.go

```diff
@@ -0,0 +1,138 @@
+	keepAlive time.Duration
+}
+
+var defaultAgentConfig = agentConfig{keepAlive: 30 * time.Second}
+
+// AgentOption customizes the behavior of an HTTP agent.
+type AgentOption func(*agentConfig)
+
+// AgentKeepAlive specifies the keep-alive period for the network connection. If
+// zero, keep-alives are disabled.
+//
+// Defaults to 30 seconds.
+func AgentKeepAlive(t time.Duration) AgentOption {
+	return func(c *agentConfig) {
+		c.keepAlive = t
+	}
+}
+
+// NewAgent creates a new http agent for managing peers and sending requests
+func NewAgent(opts ...AgentOption) *Agent {
+	cfg := defaultAgentConfig
+	for _, o := range opts {
+		o(&cfg)
+	}
+
+	return &Agent{
+		client:    buildAgentClient(&cfg),
+		peerNodes: make(map[string]*peerNode),
+	}
+}
+
+// Agent keeps track of http peers and the associated client with which the peer will call into.
+type Agent struct {
+	lock sync.RWMutex
+
+	client    *http.Client
+	peerNodes map[string]*peerNode
+}
+
+// peerNode keeps track of a HostPortPeer and any subscribers retaining it
+type peerNode struct {
+	peer        *hostport.Peer
+	subscribers map[transport.PeerSubscriber]struct{}
+}
+
+// RetainPeer gets or creates a Peer for the specificed PeerList
+func (a *Agent) RetainPeer(pid transport.PeerIdentifier, sub transport.PeerSubscriber) (transport.Peer, error) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	hppid, ok := pid.(hostport.PeerIdentifier)
+	if !ok {
+		return nil, errors.ErrInvalidPeerType{
+			ExpectedType:   "hostport.PeerIdentifier",
+			PeerIdentifier: pid,
+		}
+	}
+
+	node := a.getOrCreatePeerNode(hppid)
+	node.subscribers[sub] = struct{}{}
+	return node.peer, nil
+}
+
+// **NOTE** should only be called while the lock write mutex is acquired
+func (a *Agent) getOrCreatePeerNode(pid hostport.PeerIdentifier) *peerNode {
+	if node, ok := a.peerNodes[pid.Identifier()]; ok {
+		return node
+	}
+
+	peer := hostport.NewPeer(pid, a)
+	peer.SetStatus(transport.PeerAvailable)
+
+	node := &peerNode{
+		peer:        peer,
+		subscribers: make(map[transport.PeerSubscriber]struct{}),
+	}
+	a.peerNodes[peer.Identifier()] = node
+
+	return node
+}
+
+// ReleasePeer releases a peer from the PeerSubscriber and removes that peer from the Agent if nothing is listening to it
+func (a *Agent) ReleasePeer(pid transport.PeerIdentifier, sub transport.PeerSubscriber) error {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	node, ok := a.peerNodes[pid.Identifier()]
+	if !ok {
+		return errors.ErrAgentHasNoReferenceToPeer{
+			Agent:          a,
+			PeerIdentifier: pid,
+		}
+	}
+
+	if _, ok = node.subscribers[sub]; !ok {
+		return errors.ErrPeerHasNoReferenceToSubscriber{
+			PeerIdentifier: pid,
+			PeerSubscriber: sub,
+		}
+	}
+
+	delete(node.subscribers, sub)
+
+	if len(node.subscribers) == 0 {
+		delete(a.peerNodes, pid.Identifier())
+	}
+
+	return nil
+}
+
+// NotifyStatusChanged Notifies peer subscribers the peer's status changes.
+func (a *Agent) NotifyStatusChanged(peer transport.Peer) {
```

Comment:
this function is called twice per request, is that really reasonable to lock a mutex here?

references: https://github.com/yarpc/yarpc-go/pull/400/files#diff-4ecc4cb3ae1331409752639e243cb7cfR62
https://github.com/yarpc/yarpc-go/pull/400/files#diff-4ecc4cb3ae1331409752639e243cb7cfR70

of course optimization can always come later, i am just saying.

Additionally, I am not sure to understand why there could be more than one subscriber per peer, who are the subscribers?